Sources/GRPCHTTP2Core/Client/Resolver/DNSResolver.swift

```diff
@@ -0,0 +1,260 @@
+#endif
+@available(macOS 10.15, iOS 13.0, watchOS 6.0, tvOS 13.0, *)
+/// An asynchronous non-blocking DNS resolver built on top of the libc `getaddrinfo` function.
+package enum DNSResolver {
+  private static let dispatchQueue = DispatchQueue(
+    label: "io.grpc.DNSResolver"
+  )
+
+  /// Resolves a hostname and port number to a list of socket addresses. This method is non-blocking.
+  package static func resolve(host: String, port: Int) async throws -> [SocketAddress] {
+    if Task.isCancelled {
+      return []
+    }
+
+    return try await withCheckedThrowingContinuation { continuation in
+      Self.dispatchQueue.async {
+        do {
+          let result = try Self.resolveBlocking(host: host, port: port)
+          continuation.resume(returning: result)
+        } catch {
+          continuation.resume(throwing: error)
+        }
+      }
+    }
+  }
+
+  /// Resolves a hostname and port number to a list of socket addresses.
+  ///
+  /// Calls to `getaddrinfo` are blocking and this method calls `getaddrinfo` directly. Hence, this method is also blocking.
+  private static func resolveBlocking(host: String, port: Int) throws -> [SocketAddress] {
+    var result: UnsafeMutablePointer<addrinfo>?
+    defer {
+      if let result {
+        // Release memory allocated by a successful call to getaddrinfo
+        freeaddrinfo(result)
+      }
+    }
+
+    var hints = addrinfo()
+    hints.ai_socktype = SOCK_STREAM
+    hints.ai_protocol = IPPROTO_TCP
+
+    let errorCode = getaddrinfo(host, String(port), &hints, &result)
+
+    guard errorCode == 0, let result else {
+      throw DNSResolver.GetAddrInfoError(code: errorCode)
+    }
+
+    return try Self.parseResult(result)
+  }
+
+  /// Parses the linked list of DNS results (`addrinfo`), returning an array of socket addresses.
+  private static func parseResult(
+    _ result: UnsafeMutablePointer<addrinfo>
+  ) throws -> [SocketAddress] {
+    var result = result
+    var socketAddresses = [SocketAddress]()
+
+    while true {
+      let addressBytes: UnsafeRawPointer = .init(result.pointee.ai_addr)
+
+      switch result.pointee.ai_family {
+      case AF_INET:  // IPv4 address
+        let ipv4AddressStructure = addressBytes.load(as: sockaddr_in.self)
+        try socketAddresses.append(.ipv4(.init(ipv4AddressStructure)))
+      case AF_INET6:  // IPv6 address
+        let ipv6AddressStructure = addressBytes.load(as: sockaddr_in6.self)
+        try socketAddresses.append(.ipv6(.init(ipv6AddressStructure)))
+      default:
+        ()
+      }
+
+      guard let nextResult = result.pointee.ai_next else { break }
+      result = nextResult
+    }
+
+    return socketAddresses
+  }
+
+  /// Converts an address from a network format to a presentation format using `inet_ntop`.
+  fileprivate static func convertAddressFromNetworkToPresentationFormat<T>(
+    addressPtr: UnsafePointer<T>,
+    family: CInt,
+    length: CInt
+  ) throws -> String {
```

Comment:
You don't need the generics here
```suggestion
  fileprivate static func convertAddressFromNetworkToPresentationFormat(
    addressPtr: UnsafeRawPointer,
    family: CInt,
    length: CInt
  ) throws -> String {
```

As a caller you can still pass in the `UnsafePointer<sockaddr_in>` etc. though, it will be implicitly converted.